test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageCompatibiltyValidationTests.cs

```diff
@@ -0,0 +1,215 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.Commands;
+using NuGet.Test.Utility;
+using Xunit;
+using NuGet.Packaging.Rules;
+
+namespace NuGet.Packaging.Test
+{
+    public class PackageCompatibiltyValidationTests
+    {
+        [Fact]
+        public void NoWarningRaisedWhenBuildFound()
+        {
+            //Arrange
+            var nuspecContent = "<?xml version=\"1.0\" encoding=\"utf-8\"?>" +
+"<package xmlns=\"http://schemas.microsoft.com/packaging/2010/07/nuspec.xsd\">" +
+"   <metadata>" +
+"        <id>test</id>" +
+"        <version>1.0.0</version>" +
+"        <authors>Unit Test</authors>" +
+"        <description>Sample Description</description>" +
+"        <language>en-US</language>" +
+"    <dependencies>" +
+"      <dependency id=\"System.Collections.Immutable\" version=\"4.3.0\" />" +
+"    </dependencies>" +
+"    </metadata>" +
+"</package>";
+
+            using (var testDirectory = TestDirectory.Create())
+            {
+                var nuspecPath = Path.Combine(testDirectory, "test.nuspec");
+                File.AppendAllText(nuspecPath, nuspecContent);
+
+                Directory.CreateDirectory(Path.Combine(testDirectory, "build"));
+                Directory.CreateDirectory(Path.Combine(testDirectory, "build", "random_tfm"));
+
+                var tfmStream = File.Create(Path.Combine(testDirectory, "build", "random_tfm", "test.dll"));
+                tfmStream.Dispose();
+
+                var builder = new PackageBuilder();
+                var runner = new PackCommandRunner(
+                    new PackArgs
+                    {
+                        CurrentDirectory = testDirectory,
+                        OutputDirectory = testDirectory,
+                        Path = nuspecPath,
+                        Exclude = Array.Empty<string>(),
+                        Symbols = true,
+                        Logger = NullLogger.Instance
+                    },
+                    MSBuildProjectFactory.ProjectCreator,
+                    builder);
+
+                runner.BuildPackage();
+
+                var ruleSet = RuleSet.PackageCreationRuleSet;
+                var nupkgPath = Path.Combine(testDirectory, "test.1.0.0.nupkg");
+
+                using (var reader = new PackageArchiveReader(nupkgPath))
+                {
+                    var issues = new List<PackagingLogMessage>();
+                    foreach (var rule in ruleSet)
+                    {
+                        issues.AddRange(rule.Validate(reader).OrderBy(p => p.Code.ToString(), StringComparer.CurrentCulture));
```

Comment:
There's a lot of code to set up the tests. About 50 to set up, between 1 to 5 lines to run the test, depending how you count, and 1 assertion. Since the code being tested only uses two methods on `PackageArchiveReader`, and both are virtual, you should be able to a mocking framework, which might make the tests shorter and therefore easier to understand. Something like this:

```csharp
var mockReader = new Mock<PackageArchiveReader>();
mockReader.Setup(m => m.GetAllFiles())
  .Returns(new []
    {
      "build/random_tfm/test.dll"
    });

var target = new NoRefOrLibFolderInPackageRule();
var result = target.Validate(mockReader.Object).ToList();

Assert.Equal(0, result.Count);
```

It will also need to setup the NuspecReader and unfortunately the NuspecReader class is not mockable, but it does have a constructor that just takes a `XDocument`, so you can have a helper method to create one so your test methods are still short.

Reply:
Actually, I think NuspecReader can be mocked. The validation rule only uses GetId() which is virtual, so that should make testing the validation rules very easy.